### PR TITLE
[ImageS2S] Spatial Features Support in ImageSeq2Seq Model

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
- Add support for spatial, multidimensional features in the image-seq2seq model. 
- Add a flag, `--n-image-channels`, to specify the total spatial dimension of the image features (so that this may be combined with `--n-image-tokens`, if necessary.

I've also moved the `n_image_tokens` argument from the `ImagePolyencoderAgent` to the `TorchImageAgent`.

**Testing steps**
Added some tests to the existing `image_seq2seq` tests.

**Logs**
```
$ python -m pytest tests/test_image_seq2seq.py
================================================================================ test session starts ================================================================================
platform linux -- Python 3.6.9, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
rootdir: /private/home/kshuster/ParlAI, inifile: pytest.ini
plugins: requests-mock-1.7.0
collected 7 items

tests/test_image_seq2seq.py .......                                                                                                                                           [100%]

============================================================================= slowest 10 test durations =============================================================================
24.71s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_multitask
24.36s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_multitask_early_fusion
8.38s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_image_task
7.04s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_text_task
4.53s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_image_task_spatial_features
2.61s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_image_task_early_fusion
2.22s call     tests/test_image_seq2seq.py::TestImageSeq2Seq::test_image_task_spatial_features_early_fusion

(0.00 durations hidden.  Use -vv to show these durations.)
==================================================================== 7 passed, 198 warnings in 75.22s (0:01:15) =====================================================================
```

### CODE REVIEW

File: tests/test_image_seq2seq.py [hunk 1]
```diff
@@ -142,6 +149,43 @@ def test_multitask_early_fusion(self):
             valid['ppl'], 5.0, 'failed to train image_seq2seq on image+text task'
         )
 
+    @testing_utils.retry(ntries=3)
+    @testing_utils.skipUnlessGPU
+    def test_image_task_spatial_features(self):
+        """
+        Test that model correctly handles image task.
+
+        No training, only eval
```
Comment (EricMichaelSmith) [inline tests/test_image_seq2seq.py:158]: Hmm - wouldn't it be training because you're calling `testing_utils.train_model()`?

File: parlai/agents/image_seq2seq/modules.py [hunk 2]
```diff
@@ -294,13 +303,20 @@ def _forward_early_fusion(
...
-            )
+            # sometimes can just be a list of None
+            valid_imgs = [v for v in image_features if isinstance(v, torch.Tensor)]
+            if valid_imgs:
+                image_tensor, image_mask = self.encode_images(
+                    image_features,
+                    segments=torch.ones(  # type: ignore
+                        (
+                            len(image_features),
+                            self.n_image_channels * self.n_image_tokens,
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/agents/image_seq2seq/modules.py:314]: yeah so i think this was actually broken before for any `n_image_tokens > 1`; however, this was fine, as `n_image_tokens` could never be more than 1 for `image_seq2seq` (note that [originally this flag was not even passed in the encoder initialization](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/image_seq2seq/modules.py#L57)). indeed it was silently broadcasting to dim 1.

Segments being n_image_tokens * n_image_channels is expected, as both parameters control unfolding the image vectors into a "sequence"

File: tests/test_image_seq2seq.py (hunk 1 above)
Comment (klshuster) [reply tests/test_image_seq2seq.py:158]: ha yes, actually need to change the docstring in an above test as well
